### PR TITLE
Handle moved messages in MessageStore and message lists

### PR DESCRIPTION
**Pull request by gnprice**
Fixes most of #150. (After this lands we should do a quick scan to identify anything else that needs updating on message moves — there are several new data structures in the store that we've merged in recent weeks.)

A couple of months ago I spent a few hours writing part of an implementation of #150. It's mostly complete, apart from the lack of tests… and any bugs it has that will be found by writing those tests.

@PIG208 since you're picking up #150 shortly, I spent a small amount of time just now cleaning up that draft branch and rebasing atop the changes from #750, and writing down what remains to be done to finish it. So please start from this, which will let you avoid duplicating that work.

I believe there are three remaining things this needs:

* Tests. As usual, these should go in the same commit that makes the changes.

  There's a variety of kinds of test cases this needs,
  but here's one in particular that may not be apparent:

  If the `generation += 1` line is commented out, the message list
  has a race bug where a fetchOlder starts; we reset (because
  messages were moved into the narrow); and then the fetch returns
  and appends in the wrong spot.  Also a related race bug where it's
  fetchInitial, and it lacks the moved messages.  So there should be
  test cases exercising each of those situations, which would fail if
  that line gets commented out.

* Update StreamMessage.displayRecipient.  This goes in a followup commit.

    If we wanted, we could update it by looking up the new stream's
    name.  To enable MessageStoreImpl to do that, we'd pass it a
    StreamStore, the same way we do with Unreads.

    But in fact the only time displayRecipient is useful is when we
    don't have data on that stream.  So there's not much point in
    looking it up here.  Instead, make the field nullable, and just
    set it to null when the message is moved between streams.

* Handle UpdateMessageEvent.propagateMode.  This also goes in a followup commit. If it gets complex (or if the rest of the branch before it gets complex), let's split it out as a followup PR.

    This probably means
    _MessageListPageState having its own notion of the current narrow,
    which is initialized with `widget.narrow` but can change.
    Then `_MessageListState._modelChanged` can check if the model's
    narrow has changed, and if so tell the page state to update.

    For comparison, see zulip-mobile's `src/events/doEventActionSideEffects.js`,
    or web.

### CODE REVIEW

**Comment by PIG208** [comment]
Thanks for the notes. I think I will just push to your branch and continue using PR.

**Comment by PIG208** [comment]
Added the test cases.
This is still failing because we always notifies the listener if there are affected message present, and message moves might also notify the listener. I think we want to let the caller handle notifications instead.

File: lib/model/message_list.dart [hunk 1]
```diff
@@ -424,13 +440,15 @@ class MessageListView with ChangeNotifier, _MessageSequence {
...
     try {
+      final generation = this.generation;
       final result = await getMessages(store.connection,
         narrow: narrow.apiEncode(),
         anchor: NumericAnchor(messages[0].id),
         includeAnchor: false,
         numBefore: kMessageListFetchBatchSize,
         numAfter: 0,
       );
+      if (this.generation > generation) return;
```
Comment (PIG208) [inline lib/model/message_list.dart:451]: Not sure if we want to skip the finally block as well. At least we shouldn't need to notify the listeners again.
```
    } finally {
      _fetchingOlder = false;
      _fetchingOlder = false;
      _updateEndMarkers();
      _updateEndMarkers();
      notifyListeners();
      notifyListeners();
    }
```

**Comment by gnprice** [comment]
> This is still failing because we always notifies the listener if there are affected message present, and message moves might also notify the listener.

Hmm I see, so the listeners end up getting notified twice?

I think this is probably a case where we should just be OK with the duplication. Here's what I wrote on a different PR a few weeks ago (https://github.com/zulip/zulip-flutter/pull/648#discussion_r1623018764):
> Generally I think it's OK to have a little duplication of those calls — for example if some event can cause it to be called two or three times, and it'd take a lot of refactoring or bookkeeping to prevent that and call just once, then that's OK. If some listener callback is expensive enough to make that a problem, then it should have appropriate memoization, or defer the real work until the next frame starts, or something, in order to mitigate that.
>
> But this [in that other PR] is potentially hundreds of times, so that gets me a bit concerned.

(Possibly I should put that in docs somewhere. The way I found it just now is by a Zulip search within `#mobile-github` — that's been a side benefit I've quite enjoyed from having that GitHub firehose appear in Zulip.)

File: lib/model/message_list.dart (hunk 1 above)
Comment (gnprice) [reply lib/model/message_list.dart:451]: In the current revision:
```diff
     } finally {
       _fetchingOlder = false;
-      _updateEndMarkers();
-      notifyListeners();
+      if (this.generation == generation) {
+        _updateEndMarkers();
+        notifyListeners();
+      }
     }
```

The thing is that the reason there's a `notifyListeners` there, and an `_updateEndMarkers`, is as a followup to the `_fetchingOlder` update. Changing `_fetchingOlder` changes what set of end markers should be present; and when the end markers change, listeners should be notified (so that e.g. the UI can update to reflect the change).

We could get an effect like this, because `_fetchingOlder` may already be false. But the way to do that would be with an explicit check on `_fetchingOlder`: no _updateEndMarkers etc. if we're not actually changing the value.

In fact, this actually points to a bug (that was introduced in my draft): if we abort because the generation advanced, then we should *not* set `_fetchingOlder` to false. It might happen to be true by the time we reach this point, because the user might have scrolled up again after the new `fetchInitial` completed, causing a new `fetchOlder`. If it is true after the generation advanced, then the fact that it's true doesn't belong to this `fetchOlder` call, but instead to that newer `fetchOlder` call, and so this `fetchOlder` call shouldn't be interfering.

So please fix that bug from my draft :slightly_smiling_face:, and also add a test that would reproduce the bug.

Comment (PIG208) [reply lib/model/message_list.dart:451]: Thanks for the elaboration. I think encapsulating `_fetchingOlder` update with the generation check should be sufficient as a fix.

**Comment by PIG208** [comment]
> This probably means
_MessageListPageState having its own notion of the current narrow,
which is initialized with widget.narrow but can change.
Then _MessageListState._modelChanged can check if the model's
narrow has changed, and if so tell the page state to update.

Relying the usual updates through dependencies does not work here because `MessageListPage` is a parent of `MessageList`. At first I was thinking if we want to push a route for `MessageListPage` to `Navigator` (similar to https://github.com/zulip/zulip-mobile/issues/5251), as a way to notify `_MessageListPageState` from `_MessageListState`.

But it might not be necessary. An alternative would be having `MessageListPage` pass a callback function to `MessageList`, and call it when `_modelChanged` happens to detect change in narrow.

**Comment by gnprice** [comment]
> having `MessageListPage` pass a callback function to `MessageList`, and call it when `_modelChanged` happens to detect change in narrow.

Yeah, exactly.

We wouldn't want to push a new route — that'd leave the old one still on the nav stack, and make a new _MessageListPage State with a new _MessageListState and MessageListView. What zulip-mobile does is to mutate the route parameters on the existing route. If there's a way to do that with Flutter's navigator API then that could also be a good solution; I haven't looked into that.

File: test/model/message_list_test.dart [hunk 2]
```diff
@@ -424,6 +425,355 @@ void main() {
...
+      final narrow = TopicNarrow(stream.streamId, 'topic');
+      final initialMessages = List.generate(5, (i) => eg.streamMessage(stream: stream, topic: 'topic'));
+      final movedMessages = List.generate(5, (i) => eg.streamMessage(stream: stream, topic: 'topic'));
+
+      test('internal move', () async {
+        await prepareNarrow(narrow, initialMessages + movedMessages);
+
+        await store.handleEvent(eg.updateMessageMoveEvent(movedMessages,
+          newTopic: movedMessages[0].topic,
+          origTopic: movedMessages[0].topic
```
Comment (PIG208) [inline test/model/message_list_test.dart:459]: I wonder how this is possibly a move. Does internal move within a topic narrow actually happen?

File: lib/api/model/model.dart [hunk 3]
```diff
@@ -576,7 +576,10 @@ class StreamMessage extends Message {
   @JsonKey(includeToJson: true)
   String get type => 'stream';
 
-  final String displayRecipient;
+  // This is not nullable API-wise, but if the message move across streams we
+  // might set [displayRecipient] to null to invalidate it.
```
Comment (PIG208) [inline lib/api/model/model.dart:580]: Not sure if this comment warrants dart doc given how it is mainly explaining the motivation for the `@JsonKey` here.

File: test/model/message_list_test.dart (hunk 2 above)
Comment (gnprice) [reply test/model/message_list_test.dart:459]: Hmm. Yeah, this case seems like it should be a no-op and shouldn't happen. So maybe it should just look like this (in the implementation):
```dart
          case (true,  true ): return; // TODO(log) no-op move
```

File: lib/api/model/model.dart (hunk 3 above)
Comment (gnprice) [reply lib/api/model/model.dart:580]: Yeah, I think a plain implementation comment (like here) is fine.

**Comment by PIG208** [comment]
Pushed to fix the CI error.

**Comment by chrisbobbe** [comment]
Ah it looks like there's a rebase conflict. Could one of you please resolve that; maybe Zixuan?

**Comment by PIG208** [comment]
Conflict resolved. (It was related to 7a2c9d2d36f9b0523cf1a34d4dbbb68b7f181aca).

**Review by chrisbobbe — COMMENTED** [review]
Thanks, this will be helpful to have! Comments below.

File: test/example_data.dart [hunk 4]
```diff
@@ -465,13 +465,17 @@ UpdateMessageEvent updateMessageEditEvent(
...
+  int? origStreamId,
   int? newStreamId,
   String? origTopic,
   String? newTopic,
   String? origContent,
   String? newContent,
 }) {
   assert(messages.isNotEmpty);
+  assert(origTopic != null, 'origTopic required for a message move');
+  assert(newTopic != null, 'newTopic required for a message move');
```
[the remark this answers is not in the document]
Comment (gnprice) [reply test/example_data.dart:478]: Relatedly, if we want these asserts then they should probably just be expressed in the function's signature.

This function could also use some documentation to say how `messages` relates to the other parameters. In `updateMessageEditEvent` the Message parameter is meant to be the original, pre-edit version of the message, as reflected in its name `origMessage`. Here there's an echo of that (it's used as the default for `origStreamId`) but it doesn't seem to be applied consistently. It'd be good to have a consistent story one way or another — for example maybe it's more convenient to pass a list of the new versions of the messages, but then the function should apply that story throughout.

File: lib/model/message_list.dart [hunk 5]
```diff
@@ -485,6 +505,84 @@ class MessageListView with ChangeNotifier, _MessageSequence {
...
+  }
+
+  void _messagesMovedFromNarrow(
+    List<int> messageIds,
+    {TopicNarrow? newNarrow, PropagateMode? propagateMode}
+  ) {
+    if (_removeMessagesById(messageIds)) {
+      if (propagateMode == PropagateMode.changeAll
+          || propagateMode == PropagateMode.changeLater) {
+        narrow = newNarrow!;
```
Comment (gnprice) [inline lib/model/message_list.dart:535]: Separately: I think the overall behavior is going to be pretty broken as of this commit — the `MessageListView` changes what narrow it's tracking, but the associated `MessageListPage` is still about the old narrow. So the app bar still shows the old narrow, and the compose box will still send to the old narrow, even though the messages you're reading (including even any new messages that come in) are in the new narrow.

I think that can be fixed by just squashing the two commits together. They're both pretty small changes outside their tests.

File: test/example_data.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (PIG208) [reply test/example_data.dart:478]: > but then the function should apply that story throughout

Agreed. Will make it so that the helper always assume `messages` are the post-move messages, and express that through documentation and the removal of `newTopic` and `newStream` parameters that can then be consistently derived.

File: lib/model/message_list.dart [hunk 6]
```diff
@@ -485,6 +505,84 @@ class MessageListView with ChangeNotifier, _MessageSequence {
...
   }
 
+  void _messagesMovedInternally(List<int> messageIds) {
+    for (final messageId in messageIds) {
+      if (_findMessageWithId(messageId) != -1) {
+        _reprocessAll();
+        notifyListeners();
+        return;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (PIG208) [reply lib/model/message_list.dart:515]: Since this is returning right after an ID is found, the current implementation should be sufficient.

[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/model/message_list.dart:515]: Oh I see; yeah, I guess my eyes just skipped right past the `return`.

**Comment by PIG208** [comment]
Updated the PR. Thanks for the reviews @chrisbobbe!

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Small comments below.

I found a bug in manual testing, but it might really be a bug in the API documentation, or really the API itself, and it was latent before this PR. So maybe the fix should be a followup? Anyway, the issue is the some events are getting caught by this check in `_handleUpdateMessageEventMove`:

```dart
    if (newTopic == null) {
      // The `subject` field (aka newTopic) is documented to be present on moves.
      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic')); // TODO(log)
      return;
    }
```

It seems that when the message is moved to a new stream, but the _topic_ is the same as the topic in the original stream, the `subject` field (aka `newTopic`) is missing.

Here's the [API doc](https://zulip.com/api/get-events#update_message):

> `subject`: string
> 
> Only present if this event moved messages to a different topic.
> 
> The post-edit topic for all of the messages with IDs in `message_ids`.

**Comment by gnprice** [comment]
> I found a bug in manual testing, but it might really be a bug in the API documentation, or really the API itself,

Good find, thanks. Started an `#api documentation` thread:
https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/update_message.20events/near/1902777

File: test/example_data.dart [hunk 7]
```diff
@@ -495,6 +496,50 @@ UpdateMessageEvent updateMessageMoveEvent(
   );
 }
 
+/// [UpdateMessageEvent] where [origMessages] are moved to somewhere else.
+UpdateMessageEvent updateMessageMoveFromEvent({
```
[the remark this answers is not in the document]
Comment (gnprice) [reply test/example_data.dart:501]: Yeah, those names SGTM.

**Comment by PIG208** [comment]
Updated the PR addressing the comments. Thanks for the review! I was able to reproduce that API issue, and tweaked the checks we have to match the server behavior (also with some changes to both updateMessageEventMove helpers). Hopefully the correct behavior will documented soon.

File: test/example_data.dart [hunk 8]
```diff
@@ -495,6 +497,52 @@ UpdateMessageEvent updateMessageMoveEvent(
   );
 }
 
+/// [UpdateMessageEvent] where [origMessages] are moved to somewhere else.
+UpdateMessageEvent updateMessageEventMoveFrom({
```
Comment (chrisbobbe) [inline test/example_data.dart:502]:
```
test [nfc]: Split eg.updateMessageMoveEvent helper.
```

Is there more we can do to focus this commit on the described change? I see that `eg.updateMessageMoveEvent` gets split into two functions, so I know that that function's callers will have to adapt. But I see other changes around the callers that don't seem necessary for that adaptation, and I'm finding it difficult to verify the refactor itself. For example, it looks like a `prepareOrigMessages` helper is introduced. That seems like it should be possible to do in a separate commit.

File: lib/model/message.dart [hunk 9]
```diff
@@ -167,9 +167,10 @@ class MessageStoreImpl with MessageStore {
       return;
     }
 
-    if (newTopic == null) {
-      // The `subject` field (aka newTopic) is documented to be present on moves.
-      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic')); // TODO(log)
+    if (newTopic == null && newStreamId == null) {
+      // At least one of `subject` (aka newTopic) or `new_stream_id` (aka
+      // newStreamId) is present documented to be present on moves.
```
Comment (chrisbobbe) [inline lib/model/message.dart:173]: typo in "present documented to be present"

File: test/model/message_list_test.dart [hunk 10]
```diff
@@ -465,6 +466,337 @@ void main() {
...
+        await store.handleEvent(eg.updateMessageEventMoveFrom(
+          origMessages: movedMessages,
+          newTopic: 'new',
+        ));
+        checkHasMessages(initialMessages + movedMessages);
+        checkNotified(count: 2);
+      });
+    });
+
+    group('fetch races', (){
```
Comment (chrisbobbe) [inline test/model/message_list_test.dart:654]: nit: space before `{`

File: lib/model/message.dart [hunk 11]
```diff
@@ -167,33 +168,61 @@ class MessageStoreImpl with MessageStore {
...
+      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic or newStreamId')); // TODO(log)
```
Comment (chrisbobbe) [inline lib/model/message.dart:174]:
```
message: Handle moved messages from UpdateMessageEvent
```

Let's put this change in a separate commit before this one. It looks like it fixes a small but observable bug: in the case where you move a message to a different stream but keep the topic the same, we'll now update `message.editState` as appropriate for a move. 🙂

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Comments below.

File: test/example_data.dart (hunk 8 above)
Comment (PIG208) [reply test/example_data.dart:502]: Thanks for the advice. Extracted this part into `test [nfc]: Extract prepareOrigMessages`.

**Comment by PIG208** [comment]
Updated the PR. Thanks!

**Comment by chrisbobbe** [comment]
Thanks, LGTM! Marking for Greg's review.

**Comment by PIG208** [comment]
(pushed to resolve merge conflicts)

File: lib/model/message.dart [hunk 12]
```diff
@@ -167,33 +168,61 @@ class MessageStoreImpl with MessageStore {
...
+      // The `propagate_mode` field (aka propagateMode) is documented to be present on moves.
+      assert(debugLog('Malformed UpdateMessageEvent: move but no propagateMode')); // TODO(log)
       return;
     }
 
-    // TODO(#150): Handle message moves.  The views' recipient headers
-    //   may need updating, and consequently showSender too.
-    //   Currently only editState gets updated.
+    final wasResolveOrUnresolve = (newStreamId == null && newTopic != null
+      && MessageEditState.topicMoveWasResolveOrUnresolve(origTopic, newTopic));
```
Comment (gnprice) [inline lib/model/message.dart:189]:
```suggestion
    final wasResolveOrUnresolve = (newStreamId == null
      && MessageEditState.topicMoveWasResolveOrUnresolve(origTopic, newTopic!));
```

The case where that `newTopic != null` condition would evaluate to false should be impossible, right? That's covered by the conditional above which leads to an early return.

When a condition is meant to be impossible, it's best to make that clear rather than just not do anything for that condition. Otherwise the reader has to wonder whether there's some handling that's missing for that condition.

File: test/model/message_test.dart [hunk 13]
```diff
@@ -273,75 +273,122 @@ void main() {
...
+          newStreamId: 20,
           newContent: 'new content'));
-        check(store).messages[message.id].editState.equals(MessageEditState.edited);
-        check(store).messages[otherMessage.id].editState.equals(MessageEditState.moved);
+        checkNotified(count: 2);
+        check(store).messages[origMessages[0].id].editState.equals(MessageEditState.edited);
+        check(store).messages[origMessages[1].id].editState.equals(MessageEditState.moved);
+      });
+
+      test('message stream moved without topic change', () async {
```
Comment (gnprice) [inline test/model/message_test.dart:384]: Looks like there's two tests with this name. Is it accurate for both? If so it should still get some extra words to distinguish what each test is about.

File: test/model/message_test.dart [hunk 14]
```diff
@@ -273,75 +273,122 @@ void main() {
...
       test('message is both moved and updated', () async {
-        await sendEvent(message, eg.updateMessageMoveEvent([message, otherMessage],
-          origTopic: 'topic',
-          newTopic: 'topic',
-          newStreamId:  20,
-          origContent: 'old content',
+        await prepareOrigMessages(origTopic: 'topic', origContent: 'old content');
+        await store.handleEvent(eg.updateMessageEventMoveFrom(
+          origMessages: origMessages,
+          newTopic:  'new topic',
```
Comment (gnprice) [inline test/model/message_test.dart:376]: This `newTopic` line changes in the commit:

    model: Expect the absence of newTopic for certain moves.

Are those changes connected? Does the new topic here need to change at all — could it just remain the way the test was, where the topic doesn't get renamed but the stream ID changes?

File: lib/model/message.dart [hunk 15]
```diff
@@ -167,33 +168,61 @@ class MessageStoreImpl with MessageStore {
...
+      }
+
+      if (!wasResolveOrUnresolve
+          && message.editState == MessageEditState.none) {
+        message.editState = MessageEditState.moved;
+      }
+    }
+
+    for (final view in _messageListViews) {
+      view.messagesMoved(
```
Comment (gnprice) [inline lib/model/message.dart:218]: nit in commit message:

    Signed-off-by: Zixuan James Li <zixuan@zulip.com>
    Co-authored-by: Zixuan James Li <zixuan@zulip.com>

If we're going to have "Signed-off-by:" lines, let's follow the conventions of how they're used. It's a trailer that comes after the material you're signing off on. Are you signing off on the parts of this commit which you co-authored, or only on other parts? If the former, your "Signed-off-by:" should come after your "Co-authored-by:". (And if the latter, it's not clear what the line could really mean and it's probably best to just leave it out.)

For example usage, go into the Linux kernel repo and try a command like `git log --grep Co-author`.

**Review by gnprice — COMMENTED** [review]
Thanks @PIG208 and @chrisbobbe! Comments below.

I got through the first four commits:
c7edec418 test [nfc]: Extract prepareOrigMessages.
6d5cd3a67 test [nfc]: Split eg.updateMessageMoveEvent helper.
57a3f2099 model [nfc]: Make displayRecipient nullable.
908b7c238 model: Expect the absence of newTopic for certain moves.

and part of the fifth. Left for a future round are the remainder of the fifth and the to commits after that:
4522cfa09 message: Handle moved messages from UpdateMessageEvent.
8afbde38b msglist [nfc]: Make narrow mutable for _MessageListPageState.
b84540fda msglist: Update narrow when propageMode is changeAll or changeLater.

File: test/widgets/message_list_test.dart [hunk 16]
```diff
@@ -687,8 +688,8 @@ void main() {
       final rectsBefore = captureMessageRects(tester, messages, messageWithMarker);
       checkMarkersCount(edited: 0, moved: 0);
 
-      await store.handleEvent(eg.updateMessageMoveEvent(
-        [messageWithMarker], origTopic: 'old', newTopic: messageWithMarker.topic));
+      await store.handleEvent(eg.updateMessageEventMoveTo(
+        origTopic: 'old', newMessages: [messageWithMarker]));
```
[the remark this answers is not in the document]
Comment (PIG208) [reply test/widgets/message_list_test.dart:692]: Hmm yeah I agree. This test needs to be refactored such that messageWithMarker is initialized in another topic, before being moved into the shared one.

File: lib/model/message.dart [hunk 17]
```diff
@@ -167,33 +168,61 @@ class MessageStoreImpl with MessageStore {
       return;
     }
 
-    if (newTopic == null) {
-      // The `subject` field (aka newTopic) is documented to be present on moves.
-      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic')); // TODO(log)
+    if (newTopic == null && newStreamId == null) {
+      // At least one of `subject` (aka newTopic) or `new_stream_id` (aka
+      // newStreamId) is documented to be present on moves.
```
[the remark this answers is not in the document]
Comment (PIG208) [reply lib/model/message.dart:173]: Thanks! To summarize, prior the comment merely was an isolated interpretation of the API docs, but after the edit, it explains the assertion with context.

File: lib/model/message.dart (hunk 12 above)
Comment (PIG208) [reply lib/model/message.dart:189]: That's a good catch. Thanks!

File: test/model/message_test.dart (hunk 13 above)
Comment (PIG208) [reply test/model/message_test.dart:384]: Looks like one of them is an artifact of a misapplied fixup commit. Neither is needed, so both are removed. Renaming an old test that covers this same story.

File: test/model/message_test.dart (hunk 14 above)
Comment (PIG208) [reply test/model/message_test.dart:376]: Right, the tests changes shouldn't be necessary.

File: test/widgets/message_list_test.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (PIG208) [reply test/widgets/message_list_test.dart]: The refactor actually requires moving the message back and forth, and it has to be implemented after `message: Handle moved messages from UpdateMessageEvent.`. It wouldn't work otherwise because the move might violate the assertions (since neither topic/streamId can be updated at the time the example data helper is split).

File: test/example_data.dart [hunk 18]
```diff
@@ -463,15 +463,18 @@ UpdateMessageEvent updateMessageEditEvent(
...
   int? newStreamId,
-  String? origTopic,
   String? newTopic,
   String? origContent,
   String? newContent,
+  PropagateMode propagateMode = PropagateMode.changeOne,
 }) {
   assert(messages.isNotEmpty);
+  assert(newTopic != origTopic || newStreamId != null);
   final origMessage = messages[0];
```
[the remark this answers is not in the document]
Comment (PIG208) [reply test/example_data.dart]: Make sense. Removing the variable entirely now that we have the `flags` parameter.

File: test/widgets/message_list_test.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply test/widgets/message_list_test.dart]: I see. OK, in that case then the earlier commit can just leave a comment saying yes, this data looks wrong, it'll be fixed soon.

The blocker for fixing it is basically #150, right? In that case the comment can look like `// TODO(#150): …`.

File: test/example_data.dart [hunk 19]
```diff
@@ -463,15 +463,18 @@ UpdateMessageEvent updateMessageEditEvent(
...
-UpdateMessageEvent updateMessageMoveEvent(
+UpdateMessageEvent _updateMessageMoveEvent(
   List<Message> messages, {
+  required int origStreamId,
+  required String origTopic,
   int? newStreamId,
-  String? origTopic,
   String? newTopic,
   String? origContent,
   String? newContent,
```
[the remark this answers is not in the document]
Comment (PIG208) [reply test/example_data.dart]: While `UpdateMessageEvent` orders `flags` and `propagateMode` differently, I put these two parameters to the last for this helper to have all the orig/new pairs grouped together.

File: test/example_data.dart [hunk 20]
```diff
@@ -495,6 +498,54 @@ UpdateMessageEvent updateMessageMoveEvent(
   );
 }
 
+/// [UpdateMessageEvent] where [origMessages] are moved to somewhere else.
+UpdateMessageEvent updateMessageEventMoveFrom({
+  required List<StreamMessage> origMessages,
+  String? newTopic,
+  int? newStreamId,
```
[the remark this answers is not in the document]
Comment (PIG208) [reply test/example_data.dart]: Speaking of https://github.com/zulip/zulip-flutter/pull/787#discussion_r1696096924, do you think we should insert `propagateMode` and `flags` in between? Currently they are added to the end of the named parameters.

File: test/widgets/message_list_test.dart (hunk 16 above)
[the remark this answers is not in the document]
Comment (PIG208) [reply test/widgets/message_list_test.dart]: Yeah. This is fixed in `message_list_test: Move marker message back and forth`, after move handling is implemented. I will include the TODO comment in the refactor that touches but not fixes this test.

File: test/example_data.dart (hunk 20 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply test/example_data.dart]: I think at the end is fine — it seems like a logical place for those two. (`flags` is closely tied to `content`, which is at the end.)

Putting them in the middle, in the same order as on the event class's declaration, would also be fine. There's a bit of a tension because, ultimately, the order in the docs doesn't match a logical order; on the class we choose to match the docs.

**Comment by PIG208** [comment]
This is ready for review now. Thanks for the review!

**Review by gnprice — COMMENTED** [review]
Thanks! Those first five commits look good now:
e5576ba83 model: Expect the absence of newTopic for certain moves.
f3d00a0fc message test [nfc]: Extract prepareOrigMessages.
11f3b564d message_test [nfc]: Rename message move test for accuracy.
c3c69bce6 test [nfc]: Split eg.updateMessageMoveEvent helper.
bcd6f9688 model [nfc]: Make displayRecipient nullable.

Some comments below on the tests in the main commit:
0b54bb7e1 message: Handle moved messages from UpdateMessageEvent.

Heading into a meeting now but I might return later today to review more of that commit. Either way, there's more tests to write here, so I think you can productively start on revisions before my next review.

**Comment by PIG208** [comment]
Addressed the comments other than the new tests one. Added some additional wip tests in the last commit, and will continue working on it when I get back.

**Comment by PIG208** [comment]
Updated the tests to be more comprehensive (especially for topic narrow).

File: test/widgets/message_list_test.dart [hunk 21]
```diff
@@ -605,6 +606,86 @@ void main() {
           expectedMessage: 'NetworkException: Oops (ClientException: Oops)');
       });
     });
+
+    group('Update Narrow on message move', () {
+      final channel = eg.stream(name: 'move test stream');
+      final message = eg.streamMessage(stream: channel, content: 'Message to move');
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:612]: The message gets mutated. So each test should make it fresh — else it gets the version mutated by the last test.

File: test/widgets/message_list_test.dart [hunk 22]
```diff
@@ -605,6 +606,86 @@ void main() {
...
+
+        connection.prepare(json: SendMessageResult(id: 1).toJson());
+        await tester.tap(find.byIcon(Icons.send));
+        await tester.pump();
+        check(connection.lastRequest).isA<http.Request>()
+          ..method.equals('POST')
+          ..url.path.equals('/api/v1/messages')
+          ..bodyFields.deepEquals({
+            'type': 'stream',
+            'to': '${message.streamId}',
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:650]: Similarly, because the message gets mutated this should refer to the stream/channel object and not the message.

File: test/widgets/message_list_test.dart [hunk 23]
```diff
@@ -605,6 +606,86 @@ void main() {
...
+      }
+
+      void handleMessageMoveEvent(List<StreamMessage> messages, String newTopic) {
+        store.handleEvent(eg.updateMessageEventMoveFrom(
+          origMessages: messages,
+          newTopic: newTopic,
+          propagateMode: PropagateMode.changeAll));
+      }
+
+      testWidgets('compose box send message after move', (WidgetTester tester) async {
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:627]: Let's have this test move the message to a new channel as well as topic. That way we more fully exercise things changing.

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! Here's a review of the whole rest of the PR.

Most comments are small; there's one more-significant point about the behavior of propagateMode.

File: lib/model/message_list.dart [hunk 24]
```diff
@@ -487,6 +507,92 @@ class MessageListView with ChangeNotifier, _MessageSequence {
...
+  }
+
+  void _messagesMovedFromNarrow(
+    List<int> messageIds,
+    PropagateMode propagateMode,
+    {TopicNarrow? newNarrow}
+  ) {
+    if (_removeMessagesById(messageIds)) {
+      if (newNarrow != null) {
+        _handlePropagateMode(propagateMode, newNarrow);
```
[the remark this answers is not in the document]
Comment (PIG208) [reply lib/model/message_list.dart]: > Fixing that would be complex and is out of scope at present, even if we were to decide it's desirable.

Starting a [conversation](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Accurately.20handle.20propagateMode.2EchangeLater) so that we have a place for future discussion on this.

**Comment by PIG208** [comment]
This should be ready for review now. Thanks!

**Comment by PIG208** [comment]
(To keep track of the status) this PR is ready for review.

File: test/widgets/message_list_test.dart [hunk 25]
```diff
@@ -620,6 +621,93 @@ void main() {
     });
   });
 
+  group('Update Narrow on message move', () {
+    final channel = eg.stream(name: 'move test stream');
+    final otherChannel = eg.stream(name: 'other move test stream');
+    final narrow = TopicNarrow(channel.streamId, 'example topic');
```
[the remark this answers is not in the document]
Comment (gnprice) [reply test/widgets/message_list_test.dart:627]: > A straightforward way to do that is to make a local variable `topic` and use it in all the places that need to be talking about the same topic.

The other easy way that comes to mind is to not specify the topic name at all, but get it from the message with `message.topic` (and/or indirect versions of the same, like `TopicNarrow.ofMessage`).

File: test/widgets/message_list_test.dart [hunk 26]
```diff
@@ -620,6 +621,93 @@ void main() {
...
+      await setupMessageListPage(tester, narrow: narrow, messages: [message], streams: [channel, otherChannel]);
+
+      final channelContentInputFinder = find.descendant(
+        of: find.byType(ComposeAutocomplete),
+        matching: find.byType(TextField));
+
+      await tester.enterText(channelContentInputFinder, 'Some text');
+      check(tester.widget<TextField>(channelContentInputFinder))
+        ..decoration.isNotNull().hintText.equals('Message #${channel.name} > example topic')
+        ..controller.isNotNull().text.equals('Some text');
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:653]: Cool, good idea adding this check — it makes a baseline for the reader to compare the check below to.

File: test/widgets/message_list_test.dart [hunk 27]
```diff
@@ -620,6 +621,93 @@ void main() {
...
+      prepareGetMessageResponse([message]);
```
Comment (gnprice) [inline test/widgets/message_list_test.dart:654]: nit:
```suggestion
        ..controller.isNotNull().text.equals('Some text');

      prepareGetMessageResponse([message]);
```

Generally in a test case when there's a sequence like
* make something happen or change
* check the state of the world
* make something else happen or change
* check the new state of the world
* etc.

I find it helpful, for the reader, to group those change/check pairs into stanzas using blank lines.

File: lib/model/message_list.dart [hunk 28]
```diff
@@ -489,6 +509,87 @@ class MessageListView with ChangeNotifier, _MessageSequence {
...
+      case PropagateMode.changeAll:
+      case PropagateMode.changeLater:
+        narrow = newNarrow;
+        _reset();
+        fetchInitial();
+      case PropagateMode.changeOne:
+    }
+  }
+
+  void _messagesMovedFromNarrow(List<int> messageIds) {
```
Comment (gnprice) [inline lib/model/message_list.dart:543]: nit: group this with the other `_messagesMoved…` methods, above `_handlePropagateMode`

**Review by gnprice — COMMENTED** [review]
Thanks @PIG208 for the revision! Just a handful of small comments now; then this will be all ready to merge.

**Comment by PIG208** [comment]
Thanks. Updated the PR.

**Comment by gnprice** [comment]
Thanks for the revision! Looks good — merging.

I made a few commit-message tweaks, all in the summary prefixes:

    -    model: Expect the absence of newTopic for certain moves.
    +    message: Expect the absence of newTopic for certain moves.

    -    message_test [nfc]: Rename message move test for accuracy.
    +    message test [nfc]: Rename message move test for accuracy.

    -    model [nfc]: Make displayRecipient nullable.
    +    message [nfc]: Make displayRecipient nullable.

The "test" modifier in a summary prefix is always separated by a space, not underscore.

Rather than "model", these can be more specific by saying "message", referring to the MessageStore part of the model. (The first commit is all about that code directly. The later displayRecipient commit doesn't touch that code itself… but the change it's making to Message can be thought of as being entirely for the purpose of Message's role in MessageStore, where those objects get mutated.)

File: test/widgets/message_list_test.dart (hunk 25 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply test/widgets/message_list_test.dart]: > Probably we should make `eg.streamMessage` generate the topic name at random, to help prevent this pattern, the same way we do stream/channel IDs and message IDs and so on.

Sent #932 for this.

> and none of those assume _what_ the default topic name is, only that it's some constant value, which is also an assumption to avoid making implicit but is a different one.

I ended up deciding this assumption was fine, though; it's convenient in a variety of tests. So I made the randomness happen only once per run, and documented that.
